compute_endpoint/globus_compute_endpoint/engines/high_throughput/zmq_pipes.py

```diff
@@ -10,6 +10,13 @@
 log = logging.getLogger(__name__)
 
 
+def remap_ipv6_loopback(ip_address):
+    # Special case for being compatible with ipv4 and v6
+    if ip_address == "::1":
+        ip_address = "localhost"
+    return f"tcp://{ip_address}"
```

Comment:
I'm currently thinking we should rewrite the IP address for ZMQ's semantics.  Consider:
```python
def _zmq_canonicalize_address(addr: str | int) -> str:
    try:
        ip = ipaddress.ip_address(addr)
    except ValueError:
        # Not a valid IPv4 or IPv6 address
        if isinstance(addr, int):
            # If it was an integer, then it's just plain invalid
            raise

        # Otherwise, it was likely a hostname; let another layer deal with it
        return addr

    if ip.version == 4:
        return str(ip)  # like "12.34.56.78"
    elif ip.version == 6:
        return f"[{ip}]"  # like "[::1]"
```

This should handle all IP addresses, not just specifically `::1`.  At which point, could modify the incoming "ip" address before use, or do it inline as you've done.

Reply:
Your codewould return "::1" -> "[::1]", which results in tcp://[::1]:port right?   As opposed to my remap_func, which changes "::1" to "localhost" which results in tcp://localhost:port

That's the exact reason I made this change in the first place - neither tcp://::1:port nor tcp://[::1]:port worked (when passed to zmq).  I think there might be a way to specify it as tcpip6://hostname:port :
https://docs.quanser.com/quarc/documentation/tcpip6_protocol.html#:~:text=The%20TCP%2FIPv6%20communications%20protocol,when%20establishing%20a%20listening%20connection. 
But that may not work with zmq (I think I tried briefly and it gave me issues).

Meanwhile, tcp://localhost does work with the minimal remapping.

Reply:
> Your codewould return "::1" -> "[::1]", which results in tcp://[::1]:port right? As opposed to my remap_func, which changes "::1" to "localhost" which results in tcp://localhost:port

That's correct.  But I'm concerned with leaving "simply" `localhost:port` as without other changes, I'm not convinced that ZMQ is binding on IPv6.  The reason the bracket syntax isn't working is likely because the ZMQ context does not have IPv6 enabled&nbsp;&mdash;&nbsp;the bracket syntax is standard IPv6, and _should_ work.  Consider:

```python
>>> import zmq
>>> ctxt = zmq.Context()
>>> sock = ctxt.Socket(zmq.SERVER)
>>> sock.bind("tcp://localhost:12345")
<SocketContext(bind='tcp://127.0.0.1:12345')>
>>> sock.close()
```
That's IPv4, and we would have been none the wiser save for seeing the `repr`esentation in the terminal.  If we're explicit with the address, it's more plain that something's not right:
```python
>>> sock = ctxt.Socket(zmq.SERVER)
>>> sock.bind("tcp://[::1]:12345")
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    sock.bind("tcp://[::1]:12345")
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "VENV/zmq/sugar/socket.py", line 311, in bind
    super().bind(addr)
    ~~~~~~~~~~~~^^^^^^
  File "_zmq.py", line 917, in zmq.backend.cython._zmq.Socket.bind
  File "_zmq.py", line 179, in zmq.backend.cython._zmq._check_rc
zmq.error.ZMQError: No such device (addr='tcp://[::1]:12345')
>>> sock.close()
```
But if we tell ZMQ to use IPv6, it works as expected:
```python
>>> sock.setsockopt(zmq.IPV6, 1)
>>> sock.bind("tcp://[::1]:12345")
<SocketContext(bind='tcp://[::1]:12345')>
>>> sock.close()
```

And the `repr`esentation suggests that it has, in fact, bound via IPv6.

So ... I am not in the thick of these codes like you are just now, but my suspicion is that IPv6 is not enabled, and that that's the more fundamental detail we need to handle gracefully.

Reply:
Friendly reminder that the `HighThroughputEngine` is deprecated.

Reply:
Friendly reminder that the `HighThroughputEngine` is deprecated.

Reply:
*nod; thank you for the reminder.  Unfortunately, deprecated doesn't mean it's gone _now_.  We have an unofficial policy of at least 6 months of deprecation before removal, and we didn't publicly deprecate it [until v2.27.0](https://globus-compute.readthedocs.io/en/latest/changelog.html#globus-compute-sdk-globus-compute-endpoint-v2-27-0) (in [mid-August](https://pypi.org/project/globus-compute-endpoint/2.27.0/)).

If we're going to be IPv6 compliant ~now, I think we need to work this through.  :disappointed:

Reply:
Used your code snippet to validate ipv4/6 -